fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1971,6 +2003,9 @@ def update_p_data(custom_output_tensor: Optional[torch.Tensor] = None) -> None:
                             alloc_storage_(p._full_param_padded, size=p_size)
                         output_tensor = p._full_param_padded
 
+                    if output_tensor.dtype != p_data.dtype:
```

Comment:
@alexeib, do you have more details on what triggers this bug?

Reply:
to be honest i forgot, but i think it had to do with having fp16 params but storing fp32 grads

Reply:
haha, no worries. We will figure it out.